django/core/mail/message.py

```diff
@@ -96,16 +96,24 @@ def sanitize_address(addr, encoding):
         nm, address = addr
         localpart, domain = address.rsplit('@', 1)
 
-    nm = Header(nm, encoding).encode()
+    address_parts = nm + localpart + domain
+    if '\n' in address_parts or '\r' in address_parts:
+        raise ValueError('Invalid address; address parts cannot contain newlines.')
+
     # Avoid UTF-8 encode, if it's possible.
+    try:
+        nm.encode('ascii')
+        nm = Header(nm).encode()
+    except UnicodeEncodeError:
+        nm = Header(nm, encoding).encode()
     try:
         localpart.encode('ascii')
     except UnicodeEncodeError:
         localpart = Header(localpart, encoding).encode()
     domain = punycode(domain)
 
-    parsed_address = Address(nm, username=localpart, domain=domain)
-    return str(parsed_address)
+    parsed_address = Address(username=localpart, domain=domain)
+    return formataddr((nm, parsed_address.addr_spec), encoding)
```

Comment:
`nm` is already encoded so it's not necessary to pass `encoding`. I will remove it.